omniledger/service/service.go

```diff
@@ -473,53 +476,66 @@ func (s *Service) loadLatestDarc(sid skipchain.SkipBlockID, dID darc.ID) (*darc.
 	return darc.NewFromProtobuf(value)
 }
 
-// createQueueWorker sets up a worker that will listen on a channel for
-// incoming requests and then create a new block every epoch.
-func (s *Service) createQueueWorker(scID skipchain.SkipBlockID, interval time.Duration) chan ClientTransaction {
-	c := make(chan ClientTransaction)
+func (s *Service) startPolling(scID skipchain.SkipBlockID, interval time.Duration) chan bool {
+	closeSignal := make(chan bool)
 	go func() {
-		ts := []ClientTransaction{}
-		to := time.After(interval)
 		for {
 			select {
-			case t := <-c:
-				ts = append(ts, t)
-				log.LLvlf2("%x: Stored transaction. Next block length: %v, New Tx: %+v", scID, len(ts), t)
-			case <-to:
-				if len(ts) > 0 {
-					log.Lvlf2("%x: New epoch and transaction-length: %d", scID, len(ts))
-					sb, err := s.db().GetLatest(s.db().GetByID(scID))
-					if err != nil {
-						panic("DB is in bad state and cannot find skipchain anymore: " + err.Error())
-					}
+			case <-time.After(interval):
+				sb, err := s.db().GetLatestByID(scID)
+				if err != nil {
+					panic("DB is in bad state and cannot find skipchain anymore: " + err.Error())
+				}
+				// We assume the caller of this function is the
+				// current leader. So we generate a tree with
+				// this node as the root.
+				tree := sb.Roster.GenerateNaryTreeWithRoot(len(sb.Roster.List), s.ServerIdentity())
+				proto, err := s.CreateProtocol(collectTxProtocol, tree)
+				if err != nil {
+					panic("cannot create protocol: " + err.Error())
```

Comment:
If someone, somehow, finds a way to make this happen, then the entire server will crash. There's got to be a better reaction than panic here., like maybe log.Error() and close(closeSignal).

Reply:
The reason I decided to panic is that there is simply nothing we can do to recover without putting the server in a strange state. If you stop the go-routine than that would result in a non-polling leader. Then somebody has to go and fix the problem and restart the server.

Reply:
ok, good reasoning. Can you put it in the code, so that later when we say, "damn it, why is this panic here?" we don't forget?

Reply:
sure, I updated it